File: src/239_Sliding_Window_Maximum.py
import heapq
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        n = len(nums)
        win = [(-nums[i], i) for i in range(k)]
        heapq.heapify(win)
        ans = [-win[0][0]]
        for i in range(k,n):
            heapq.heappush(win,(-nums[i],i))
            while win[0][1] <= i - k:
                heapq.heappop(win)
            ans.append(-win[0][0])
        return ans

File: src/239_Sliding_Window_Maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        priority_queue<pair<int,int>> win;

        for(int i = 0; i < k; i++){
            win.emplace(nums[i],i);
        }

        vector<int> ans;
        ans.push_back(win.top().first);
        for(int i = k; i < nums.size(); i++){
            win.emplace(nums[i],i);
            while(win.top().second <= i - k){
                win.pop();
            }
            ans.push_back(win.top().first);
        }
        return ans;
    }
};